torch/onnx/_internal/torchscript.py

```diff
@@ -0,0 +1,426 @@
+"""Utilities for manipulating the torch.Graph object and the torchscript."""
+
+# TODO(justinchuby): Move more of the symbolic helper functions here and expose
+# them to the user.
+
+import dataclasses
+import numbers
+import re
+from typing import Any, Dict, Iterable, Iterator, List, Tuple, Union
+
+from typing_extensions import Protocol, runtime_checkable
+
+import torch
+from torch import _C
+from torch._C import _onnx as _C_onnx
+from torch.onnx import _deprecation
+from torch.onnx._globals import GLOBALS
+from torch.onnx._internal import _beartype
+
+_ATTR_PATTERN = re.compile("^(.+)_(([ifstgz])|(ty))$")
+
+
+@runtime_checkable
+class GraphLike(Protocol):
```

Comment:
Duplicating each  method defined at https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/python_ir.cpp#L218 is error prone. Users may tap new methods to the pybind class and not change this one (because they dont know about it).

Did you consider creating a `GraphAbstract` class using `class GraphInterface(metaclass=abc.ABCMeta)` to define an Interface and then change https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/python/python_ir.cpp#L218 to implement such interface?

With that, `GraphAbstract` is the only source of truth for the public interface and the `Graph` and/or `GraphContext` implement them as needed, inheriting from the same class

Reply:
I am conscious about creating dependencies on Python code in cpp. They are usually harder to debug. This Protocol is for type checkers and ides only. If the cpp code is changed, we can expect the new method usable because getattr in GraphContext looks for any method that is not defined in class.

Reply:
I am going to remove `GraphLike` and use GraphContext for everything

Reply:
I wonder what's the life cycle design for `GraphContext`. Is it for symbolic function use only, hence created every time in `_run_symbolic_function`, or is it something that will be passed around though out entire export? If it's the latter, we should take cpp interactions into consideration.

Reply:
For now it is created every time in `_run_symbolic_function` so there is no interactions yet. Any ideas?